File: dart/lib/src/proof/im.dart
import '../codec/base64.dart';
import '../errors.dart';
import '../keys/key_material.dart';

const List<String> imProofDefaultComponents = [
  '@method',
  '@target-uri',
  'content-digest',
];
const String imProofRelationshipAuthentication = 'authentication';
const String imProofRelationshipAssertionMethod = 'assertionMethod';

class ImProof {
  const ImProof({
    required this.contentDigest,
    required this.signatureInput,
    required this.signature,
  });
  final String contentDigest;
  final String signatureInput;
  final String signature;
}

class ParsedImSignatureInput {
  const ParsedImSignatureInput({
    required this.label,
    required this.keyId,
    required this.components,
    required this.signatureParams,
    this.nonce,
    this.created,
    this.expires,
  });
  final String label;
  final String keyId;
  final List<String> components;
  final String signatureParams;
  final String? nonce;
  final int? created;
  final int? expires;
}

String buildImContentDigest(List<int> payload) =>
    'sha-256=:${encodeBase64(sha256Bytes(payload))}:';

bool verifyImContentDigest(List<int> payload, String digest) =>
    buildImContentDigest(payload) == digest;

String buildImSignatureInput(
  String keyId, {
  String label = 'sig1',
  List<String> components = imProofDefaultComponents,
  int? created,
  int? expires,
  String? nonce,
}) {
  if (keyId.trim().isEmpty) {
    throw const AnpProofException('proof.signatureInput must include keyid');
  }
  final actualLabel = label.trim().isEmpty ? 'sig1' : label.trim();
  final actualComponents = components.isEmpty
      ? imProofDefaultComponents
      : components;
  if (actualComponents.isEmpty) {
    throw const AnpProofException(
      'proof.signatureInput must include covered components',
    );
  }
  final actualCreated =
      created ?? DateTime.now().toUtc().millisecondsSinceEpoch ~/ 1000;
  final actualNonce = nonce == null || nonce.isEmpty
      ? encodeBase64Url(
          sha256Bytes(
            utf8Bytes(DateTime.now().microsecondsSinceEpoch.toString()),
          ).sublist(0, 16),
        )
      : nonce;
  final params = <String>[
    'created=$actualCreated',
    if (expires != null) 'expires=$expires',
    'nonce="$actualNonce"',
    'keyid="$keyId"',
  ];
  return '$actualLabel=(${actualComponents.map((value) => '"$value"').join(' ')});${params.join(';')}';
}

ParsedImSignatureInput parseImSignatureInput(String value) {
  final separator = value.indexOf('=');
  if (separator < 0) {
    throw const AnpProofException('invalid proof.signatureInput format');
  }
  final label = value.substring(0, separator).trim();
  final remainder = value.substring(separator + 1).trim();
  final openIndex = remainder.indexOf('(');
  final closeIndex = remainder.indexOf(')');
  if (openIndex < 0 || closeIndex <= openIndex) {
    throw const AnpProofException('invalid proof.signatureInput format');
  }
  final components = remainder
      .substring(openIndex + 1, closeIndex)
      .split(RegExp(r'\s+'))
      .where((part) => part.isNotEmpty)
      .map((part) => part.replaceAll('"', ''))
      .toList();
  if (components.isEmpty) {
    throw const AnpProofException(
      'proof.signatureInput must include covered components',
    );
  }
  final signatureParams = remainder;
  final params = _parseKvParams(
    remainder.substring(closeIndex + 1).trim().replaceFirst(RegExp(r'^;'), ''),
  );
  final keyId = params['keyid'] ?? '';
  if (keyId.isEmpty) {
    throw const AnpProofException('proof.signatureInput must include keyid');
  }
  return ParsedImSignatureInput(
    label: label,
    keyId: keyId,
    components: components,
    signatureParams: signatureParams,
    nonce: params['nonce'],
    created: int.tryParse(params['created'] ?? ''),
    expires: int.tryParse(params['expires'] ?? ''),
  );
}

String encodeImSignature(List<int> signature, {String label = 'sig1'}) =>
    '$label=:${encodeBase64(signature)}:';

({String label, List<int> signature}) decodeImSignature(String signature) {
  final trimmed = signature.trim();
  final separator = trimmed.indexOf('=:');
  if (separator >= 0) {
    final label = trimmed.substring(0, separator);
    final encoded = trimmed.substring(separator + 2);
    if (!encoded.endsWith(':')) {
      throw const AnpProofException('invalid proof.signature encoding');
    }
    return (
      label: label,
      signature: decodeBase64(encoded.substring(0, encoded.length - 1)),
    );
  }
  final bare = trimmed.replaceAll(RegExp(r'^:+|:+$'), '');
  return (label: '', signature: decodeBase64(bare));
}

Map<String, String> _parseKvParams(String value) {
  final result = <String, String>{};
  for (final part in value.split(';')) {
    final trimmed = part.trim();
    if (trimmed.isEmpty) continue;
    final separator = trimmed.indexOf('=');
    if (separator < 0) continue;
    final key = trimmed.substring(0, separator).trim();
    var rawValue = trimmed.substring(separator + 1).trim();
    if (rawValue.startsWith('"') && rawValue.endsWith('"')) {
      rawValue = rawValue.substring(1, rawValue.length - 1);
    }
    result[key] = rawValue;
  }
  return result;
}

File: golang/proof/im.go
package proof

import (
	"crypto/sha256"
	"fmt"
	"strconv"
	"strings"
	"time"

	anp "github.com/agent-network-protocol/anp/golang"
	"github.com/agent-network-protocol/anp/golang/internal/base64util"
	"github.com/agent-network-protocol/anp/golang/internal/cjson"
	"github.com/agent-network-protocol/anp/golang/internal/diddoc"
	"github.com/agent-network-protocol/anp/golang/internal/vm"
)

var IMProofDefaultComponents = []string{"@method", "@target-uri", "content-digest"}

const (
	// IMProofRelationshipAuthentication binds request-level IM proofs to DID authentication.
	IMProofRelationshipAuthentication = "authentication"
	// IMProofRelationshipAssertionMethod binds assertion-style IM proofs to DID assertionMethod.
	IMProofRelationshipAssertionMethod = "assertionMethod"
)

// IMProof stores an ANP IM business proof.
type IMProof struct {
	ContentDigest  string `json:"contentDigest"`
	SignatureInput string `json:"signatureInput"`
	Signature      string `json:"signature"`
}

// ParsedIMSignatureInput is the parsed signature-input structure.
type ParsedIMSignatureInput struct {
	Label           string   `json:"label"`
	Components      []string `json:"components"`
	SignatureParams string   `json:"signature_params"`
	KeyID           string   `json:"keyid"`
	Nonce           string   `json:"nonce,omitempty"`
	Created         *int64   `json:"created,omitempty"`
	Expires         *int64   `json:"expires,omitempty"`
}

// IMGenerationOptions configures IM proof generation.
type IMGenerationOptions struct {
	Label      string
	Components []string
	Created    *int64
	Expires    *int64
	Nonce      string
}

// IMVerificationResult contains the verified verification method.
type IMVerificationResult struct {
	ParsedSignatureInput ParsedIMSignatureInput `json:"parsed_signature_input"`
	VerificationMethod   map[string]any         `json:"verification_method"`
}

// BuildIMContentDigest builds a content-digest header value for IM payloads.
func BuildIMContentDigest(payload []byte) string {
	hash := sha256.Sum256(payload)
	return "sha-256=:" + base64util.EncodeStd(hash[:]) + ":"
}

// VerifyIMContentDigest verifies a content-digest value.
func VerifyIMContentDigest(payload []byte, contentDigest string) bool {
	return BuildIMContentDigest(payload) == strings.TrimSpace(contentDigest)
}

// BuildIMSignatureInput builds an IM signature-input value.
func BuildIMSignatureInput(keyID string, options IMGenerationOptions) (string, error) {
	label := options.Label
	if label == "" {
		label = "sig1"
	}
	components := options.Components
	if len(components) == 0 {
		components = append([]string(nil), IMProofDefaultComponents...)
	}
	if len(components) == 0 {
		return "", &Error{Message: "proof.signatureInput must include covered components"}
	}
	created := time.Now().Unix()
	if options.Created != nil {
		created = *options.Created
	}
	nonce := options.Nonce
	if nonce == "" {
		nonce = anp.EncodeBase64URL(randomNonce())
	}
	parts := []string{fmt.Sprintf("created=%d", created)}
	if options.Expires != nil {
		parts = append(parts, fmt.Sprintf("expires=%d", *options.Expires))
	}
	parts = append(parts, fmt.Sprintf("nonce=\"%s\"", nonce), fmt.Sprintf("keyid=\"%s\"", keyID))
	quoted := make([]string, 0, len(components))
	for _, component := range components {
		quoted = append(quoted, fmt.Sprintf("\"%s\"", component))
	}
	return fmt.Sprintf("%s=(%s);%s", label, strings.Join(quoted, " "), strings.Join(parts, ";")), nil
}

// ParseIMSignatureInput parses a signature-input string.
func ParseIMSignatureInput(signatureInput string) (ParsedIMSignatureInput, error) {
	separator := strings.IndexByte(signatureInput, '=')
	if separator < 0 {
		return ParsedIMSignatureInput{}, &Error{Message: "invalid proof.signatureInput format"}
	}
	label := strings.TrimSpace(signatureInput[:separator])
	remainder := strings.TrimSpace(signatureInput[separator+1:])
	openIndex := strings.IndexByte(remainder, '(')
	closeIndex := strings.IndexByte(remainder, ')')
	if openIndex < 0 || closeIndex <= openIndex {
		return ParsedIMSignatureInput{}, &Error{Message: "invalid proof.signatureInput format"}
	}
	componentTokens := strings.Fields(remainder[openIndex+1 : closeIndex])
	components := make([]string, 0, len(componentTokens))
	for _, token := range componentTokens {
		components = append(components, strings.Trim(token, "\""))
	}
	if len(components) == 0 {
		return ParsedIMSignatureInput{}, &Error{Message: "proof.signatureInput must include covered components"}
	}
	params := parseKVParams(strings.TrimPrefix(strings.TrimSpace(remainder[closeIndex+1:]), ";"))
	keyID := params["keyid"]
	if keyID == "" {
		return ParsedIMSignatureInput{}, &Error{Message: "proof.signatureInput must include keyid"}
	}
	result := ParsedIMSignatureInput{
		Label:           label,
		Components:      components,
		SignatureParams: remainder,
		KeyID:           keyID,
		Nonce:           params["nonce"],
	}
	if created := params["created"]; created != "" {
		if value, err := strconv.ParseInt(created, 10, 64); err == nil {
			result.Created = &value
		}
	}
	if expires := params["expires"]; expires != "" {
		if value, err := strconv.ParseInt(expires, 10, 64); err == nil {
			result.Expires = &value
		}
	}
	return result, nil
}

// EncodeIMSignature encodes signature bytes into structured signature syntax.
func EncodeIMSignature(signatureBytes []byte, label string) string {
	if label == "" {
		label = "sig1"
	}
	return fmt.Sprintf("%s=:%s:", label, base64util.EncodeStd(signatureBytes))
}

// DecodeIMSignature decodes a structured or bare signature value.
func DecodeIMSignature(signature string) (string, []byte, error) {
	trimmed := strings.TrimSpace(signature)
	if strings.Contains(trimmed, "=:") {
		parts := strings.SplitN(trimmed, "=:", 2)
		if len(parts) != 2 || !strings.HasSuffix(parts[1], ":") {
			return "", nil, &Error{Message: "invalid proof.signature encoding"}
		}
		decoded, err := base64util.DecodeStd(strings.TrimSuffix(parts[1], ":"))
		if err != nil {
			decoded, err = base64util.DecodeURL(strings.TrimSuffix(parts[1], ":"))
			if err != nil {
				return "", nil, &Error{Message: "invalid proof.signature encoding"}
			}
		}
		return parts[0], decoded, nil
	}
	bare := strings.Trim(trimmed, ":")
	decoded, err := base64util.DecodeStd(bare)
	if err != nil {
		decoded, err = base64util.DecodeURL(bare)
		if err != nil {
			return "", nil, &Error{Message: "invalid proof.signature encoding"}
		}
	}
	return "", decoded, nil
}

// GenerateIMProof signs a prepared signature base and payload digest.
func GenerateIMProof(payload []byte, signatureBase []byte, privateKey anp.PrivateKeyMaterial, keyID string, options IMGenerationOptions) (IMProof, error) {
	signatureInput, err := BuildIMSignatureInput(keyID, options)
	if err != nil {
		return IMProof{}, err
	}
	signatureBytes, err := privateKey.SignMessage(signatureBase)
	if err != nil {
		return IMProof{}, &Error{Message: "signing error"}
	}
	label := options.Label
	if label == "" {
		label = "sig1"
	}
	return IMProof{ContentDigest: BuildIMContentDigest(payload), SignatureInput: signatureInput, Signature: EncodeIMSignature(signatureBytes, label)}, nil
}

// VerifyIMProofWithDocument verifies an IM proof using a DID document.
func VerifyIMProofWithDocument(proof IMProof, payload []byte, signatureBase []byte, didDocument map[string]any, expectedSignerDID string) (IMVerificationResult, error) {
	return VerifyIMProofWithDocumentForRelationship(proof, payload, signatureBase, didDocument, expectedSignerDID, IMProofRelationshipAuthentication)
}

// VerifyIMProofWithDocumentForRelationship verifies an IM proof using a DID document and an explicit verification relationship.
func VerifyIMProofWithDocumentForRelationship(proof IMProof, payload []byte, signatureBase []byte, didDocument map[string]any, expectedSignerDID string, verificationRelationship string) (IMVerificationResult, error) {
	parsed, err := ParseIMSignatureInput(proof.SignatureInput)
	if err != nil {
		return IMVerificationResult{}, err
	}
	if expectedSignerDID != "" && !keyIDBelongsToExpectedDID(parsed.KeyID, expectedSignerDID) {
		return IMVerificationResult{}, &Error{Message: "proof keyid must belong to expected signer DID"}
	}
	relationship, err := normalizeIMVerificationRelationship(verificationRelationship)
	if err != nil {
		return IMVerificationResult{}, err
	}
	if !isIMVerificationMethodAuthorized(didDocument, parsed.KeyID, relationship) {
		return IMVerificationResult{}, &Error{Message: fmt.Sprintf("verification method is not authorized for %s", relationship)}
	}
	verificationMethod := diddoc.FindVerificationMethod(didDocument, parsed.KeyID)
	if verificationMethod == nil {
		return IMVerificationResult{}, &Error{Message: "verification method not found in DID document"}
	}
	return VerifyIMProofWithVerificationMethod(proof, payload, signatureBase, verificationMethod, expectedSignerDID)
}

// VerifyIMProofWithVerificationMethod verifies an IM proof with an explicit verification method.
func VerifyIMProofWithVerificationMethod(proof IMProof, payload []byte, signatureBase []byte, verificationMethod map[string]any, expectedSignerDID string) (IMVerificationResult, error) {
	if !VerifyIMContentDigest(payload, proof.ContentDigest) {
		return IMVerificationResult{}, &Error{Message: "proof contentDigest does not match request payload"}
	}
	parsed, err := ParseIMSignatureInput(proof.SignatureInput)
	if err != nil {
		return IMVerificationResult{}, err
	}
	if expectedSignerDID != "" && !keyIDBelongsToExpectedDID(parsed.KeyID, expectedSignerDID) {
		return IMVerificationResult{}, &Error{Message: "proof keyid must belong to expected signer DID"}
	}
	label, signatureBytes, err := DecodeIMSignature(proof.Signature)
	if err != nil {
		return IMVerificationResult{}, err
	}
	if label != "" && label != parsed.Label {
		return IMVerificationResult{}, &Error{Message: "invalid proof.signature encoding"}
	}
	parsedMethod, err := vm.CreateVerificationMethod(verificationMethod)
	if err != nil {
		return IMVerificationResult{}, &Error{Message: "verification method not found in DID document"}
	}
	if err := parsedMethod.PublicKey.VerifyMessage(signatureBase, signatureBytes); err != nil {
		return IMVerificationResult{}, &Error{Message: "signature verification failed"}
	}
	return IMVerificationResult{ParsedSignatureInput: parsed, VerificationMethod: cloneMap(verificationMethod)}, nil
}

func parseKVParams(value string) map[string]string {
	result := map[string]string{}
	for _, part := range strings.Split(value, ";") {
		trimmed := strings.TrimSpace(part)
		if trimmed == "" {
			continue
		}
		name, rawValue, ok := strings.Cut(trimmed, "=")
		if !ok {
			continue
		}
		result[strings.TrimSpace(name)] = strings.Trim(strings.TrimSpace(rawValue), "\"")
	}
	return result
}

func randomNonce() []byte {
	seed, _ := cjson.Marshal(map[string]any{"time": time.Now().UnixNano()})
	hash := sha256.Sum256(seed)
	return hash[:16]
}

func normalizeIMVerificationRelationship(verificationRelationship string) (string, error) {
	if verificationRelationship == "" {
		return IMProofRelationshipAuthentication, nil
	}
	if verificationRelationship != IMProofRelationshipAuthentication && verificationRelationship != IMProofRelationshipAssertionMethod {
		return "", &Error{Message: fmt.Sprintf("unsupported verification relationship: %s", verificationRelationship)}
	}
	return verificationRelationship, nil
}

func isIMVerificationMethodAuthorized(didDocument map[string]any, verificationMethodID string, verificationRelationship string) bool {
	switch verificationRelationship {
	case IMProofRelationshipAuthentication:
		return diddoc.IsAuthenticationAuthorized(didDocument, verificationMethodID)
	case IMProofRelationshipAssertionMethod:
		return diddoc.IsAssertionMethodAuthorized(didDocument, verificationMethodID)
	default:
		return false
	}
}

func keyIDBelongsToExpectedDID(keyID string, expectedSignerDID string) bool {
	return strings.SplitN(keyID, "#", 2)[0] == expectedSignerDID
}
